pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/SortedMailboxReceiveOperator.java

```diff
@@ -0,0 +1,202 @@
+            // Get null block when pulling times out from mailbox.
+            if (block != null) {
+              foundNonNullTransferableBlock = true;
+              if (block.isErrorBlock()) {
+                _upstreamErrorBlock =
+                    TransferableBlockUtils.getErrorTransferableBlock(block.getDataBlock().getExceptions());
+                return _upstreamErrorBlock;
+              }
+              if (!block.isEndOfStreamBlock()) {
+                // Add rows to the PriorityQueue to order them
+                List<Object[]> container = block.getContainer();
+                _priorityQueue.addAll(container);
+              } else {
+                if (_opChainStats != null && !block.getResultMetadata().isEmpty()) {
+                  for (Map.Entry<String, OperatorStats> entry : block.getResultMetadata().entrySet()) {
+                    _opChainStats.getOperatorStatsMap().compute(entry.getKey(), (_key, _value) -> entry.getValue());
+                  }
+                }
+                eosMailboxCount++;
+              }
+            }
+          }
+        } catch (Exception e) {
+          return TransferableBlockUtils.getErrorTransferableBlock(
+              new RuntimeException(String.format("Error polling mailbox=%s", mailboxId), e));
+        }
+      }
+    } while (foundNonNullTransferableBlock && ((openMailboxCount > 0) && (openMailboxCount > eosMailboxCount)));
+
+    if (((openMailboxCount == 0) || (openMailboxCount == eosMailboxCount))
+        && (!CollectionUtils.isEmpty(_priorityQueue)) && !_isSortedBlockConstructed) {
+      // Some data is present in the PriorityQueue, these need to be sent upstream
+      List<Object[]> rows = new ArrayList<>();
+      while (_priorityQueue.size() > 0) {
+        Object[] row = _priorityQueue.poll();
+        rows.add(row);
+      }
+      _isSortedBlockConstructed = true;
+      return new TransferableBlock(rows, _dataSchema, DataBlock.Type.ROW);
+    }
+
+    // there are two conditions in which we should return EOS: (1) there were
+    // no mailboxes to open (this shouldn't happen because the second condition
+    // should be hit first, but is defensive) (2) every mailbox that was opened
+    // returned an EOS block. in every other scenario, there are mailboxes that
+    // are not yet exhausted and we should wait for more data to be available
+    TransferableBlock block =
+        openMailboxCount > 0 && openMailboxCount > eosMailboxCount ? TransferableBlockUtils.getNoOpTransferableBlock()
+            : TransferableBlockUtils.getEndOfStreamTransferableBlock();
+    return block;
+  }
+
+  private void cleanUpResources() {
+    if (_priorityQueue != null) {
+      _priorityQueue.clear();
+    }
+  }
+
+  @Override
+  public void close() {
+    super.close();
+    cleanUpResources();
+  }
+
+  @Override
+  public void cancel(Throwable t) {
+    super.cancel(t);
+    if (_upstreamErrorBlock != null) {
+      cleanUpResources();
```

Comment:
I think we should call this everytime. For example in line 153 we don't set the upstream error-block, and in the future there may be a few more edge cases.

```
        } catch (Exception e) {
          return TransferableBlockUtils.getErrorTransferableBlock(
              new RuntimeException(String.format("Error polling mailbox=%s", mailboxId), e));
        }
```

Reply:
ah good point. done